magnum-cats-effect/src/main/scala/com/augustnagro/magnum/magcats/Transactor.scala

```diff
@@ -0,0 +1,202 @@
+package com.augustnagro.magnum.magcats
+
+import cats.*
+import cats.effect.kernel.*
+import cats.effect.kernel.Outcome.*
+import cats.effect.std.*
+import cats.effect.syntax.all.*
+import cats.syntax.all.*
+import com.augustnagro.magnum.*
+
+import java.sql.Connection
+import javax.sql.DataSource
+import scala.util.control.NonFatal
+
+class Transactor[F[_]: Sync] private (
+    dataSource: DataSource,
+    sqlLogger: SqlLogger,
+    connectionConfig: Connection => Unit,
+    rateLimiter: Option[Resource[F, Unit]]
+):
+  private val makeConn = Resource.make(acquireConnection)(releaseConnection)
+
+  def withSqlLogger(sqlLogger: SqlLogger): Transactor[F] =
+    new Transactor(
+      dataSource,
+      sqlLogger,
+      connectionConfig,
+      rateLimiter
+    )
+
+  def withConnectionConfig(
+      connectionConfig: Connection => Unit
+  ): Transactor[F] =
+    new Transactor(
+      dataSource,
+      sqlLogger,
+      connectionConfig,
+      rateLimiter
+    )
+
+  def connect[A](f: DbCon ?=> A): F[A] =
+    useRateLimitedConnection: cn =>
+      Sync[F].delay(connectionConfig(cn)) >>
+        Sync[F].interruptible(f(using DbCon(cn, sqlLogger)))
+
+  def transact[A](f: DbTx ?=> A): F[A] =
+    useRateLimitedConnection: cn =>
+      Sync[F].delay(connectionConfig(cn)) >>
+        Sync[F].delay(cn.setAutoCommit(false)) >>
+        Sync[F]
+          .interruptible(f(using DbTx(cn, sqlLogger)))
+          .guaranteeCase {
+            case Succeeded(_) => Sync[F].blocking(cn.commit())
+            case Errored(_) | Canceled() =>
+              Sync[F].blocking(cn.rollback())
+          }
+
+  private def useRateLimitedConnection[A](program: Connection => F[A]): F[A] =
+    val io = makeConn.use(program)
+    rateLimiter.fold(io)(_.surround(io))
+
+  private def acquireConnection: F[Connection] =
+    Sync[F]
+      .delay(dataSource.getConnection())
```

Comment:
This is potentially blocking I think
In ZIO code, the `acquireConnection` private method is not wrapping the `dataSource.getConnection()` in a blocking context because this `acquireConnection` private method is always used in a `ZIO.blocking(...)` context

Reply:
Thanks, fixed